File: OpenPose/OpenPoseImage.cpp
#include "pose_estimation.hpp"

#include <opencv2/highgui.hpp>
#include <opencv2/imgcodecs.hpp>

#include <filesystem>
#include <iostream>
#include <stdexcept>
#include <string>

#ifndef OPENPOSE_SOURCE_DIR
#define OPENPOSE_SOURCE_DIR "."
#endif

namespace {

struct Options {
    std::filesystem::path input =
        std::filesystem::path(OPENPOSE_SOURCE_DIR) / "single.jpeg";
    std::filesystem::path model =
        std::filesystem::path(OPENPOSE_SOURCE_DIR) /
        "models" /
        "pose_estimation_mediapipe_2023mar.onnx";
    std::filesystem::path output_dir =
        std::filesystem::path(OPENPOSE_SOURCE_DIR) / "output";
    std::string device = "cpu";
    float score_threshold = 0.5F;
    bool display = false;
    bool validate = false;
    bool help = false;
};

std::string requireValue(int& index, int argc, char** argv) {
    if (index + 1 >= argc) {
        throw std::invalid_argument(
            "Missing value after " + std::string(argv[index]));
    }
    ++index;
    return argv[index];
}

Options parseOptions(int argc, char** argv) {
    Options options;
    for (int index = 1; index < argc; ++index) {
        const std::string argument = argv[index];
        if (argument == "--help" || argument == "-h") {
            options.help = true;
        } else if (argument == "--input") {
            options.input = requireValue(index, argc, argv);
        } else if (argument == "--model") {
            options.model = requireValue(index, argc, argv);
        } else if (argument == "--output-dir") {
            options.output_dir = requireValue(index, argc, argv);
        } else if (argument == "--device") {
            options.device = requireValue(index, argc, argv);
        } else if (argument == "--score-threshold") {
            options.score_threshold =
                std::stof(requireValue(index, argc, argv));
        } else if (argument == "--display") {
            options.display = true;
        } else if (argument == "--no-display") {
            options.display = false;
        } else if (argument == "--validate") {
            options.validate = true;
        } else {
            throw std::invalid_argument("Unknown argument: " + argument);
        }
    }
    if (options.score_threshold < 0.0F || options.score_threshold > 1.0F) {
        throw std::invalid_argument(
            "--score-threshold must be between 0 and 1.");
    }
    return options;
}

void printHelp(const char* executable) {
    std::cout
        << "Usage: " << executable << " [options]\n"
        << "  --input PATH             Input image\n"
        << "  --model PATH             MediaPipe Pose ONNX model\n"
        << "  --output-dir PATH        Directory for pose-image.jpg\n"
        << "  --device cpu|cuda        DNN execution device\n"
        << "  --score-threshold VALUE  Landmark probability threshold\n"
        << "  --display                Open an interactive window\n"
        << "  --no-display             Run headlessly (default)\n"
        << "  --validate               Check stable output invariants\n";
}

}  // namespace

int main(int argc, char** argv) {
    try {
        const Options options = parseOptions(argc, argv);
        if (options.help) {
            printHelp(argv[0]);
            return 0;
        }

        // imread returns an empty matrix for both missing and undecodable files.
        const cv::Mat frame =
            cv::imread(options.input.string(), cv::IMREAD_COLOR);
        if (frame.empty()) {
            throw std::runtime_error(
                "Could not read input image: " + options.input.string());
        }

        learnopencv::pose::PoseEstimator estimator(
            options.model, options.device);
        const learnopencv::pose::PoseResult result = estimator.infer(frame);

        cv::Mat output = frame.clone();
        const learnopencv::pose::DrawMetrics metrics =
            learnopencv::pose::draw(
                output, result, options.score_threshold);

        std::filesystem::create_directories(options.output_dir);
        const std::filesystem::path output_path =
            options.output_dir / "pose-image.jpg";
        if (!cv::imwrite(output_path.string(), output)) {
            throw std::runtime_error(
                "Could not write output image: " + output_path.string());
        }

        if (options.validate) {
            learnopencv::pose::validate(frame, result, metrics);
            std::cout
                << "VALIDATION PASSED: landmarks=33 visible="
                << metrics.visible_count
                << " edges=" << metrics.edge_count << '\n';
        }

        if (options.display) {
            cv::imshow("MediaPipe Pose", output);
            cv::waitKey(0);
            cv::destroyAllWindows();
        }

        std::cout << "OpenCV version: " << CV_VERSION << '\n';
        std::cout
            << "POSE RESULT: confidence=" << result.confidence
            << " visible=" << metrics.visible_count
            << " edges=" << metrics.edge_count << '\n';
        std::cout << "Saved output: " << output_path << '\n';
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "ERROR: " << error.what() << '\n';
        return 2;
    }
}

File: OpenPose/OpenPoseImage.py
#!/usr/bin/env python3
"""Estimate one person's MediaPipe Pose landmarks in an image."""

from __future__ import annotations

import argparse
from pathlib import Path
import sys

import cv2

from pose_estimation import (
    DEFAULT_MODEL,
    PROJECT_DIR,
    draw_pose,
    infer_pose,
    load_pose_model,
    validate_pose,
    write_image,
)


DEFAULT_INPUT = PROJECT_DIR / "single.jpeg"


def parse_args(argv: list[str] | None = None) -> argparse.Namespace:
    """Expose reproducible input, output, device, display, and validation controls."""

    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--input",
        "--image_file",
        dest="input",
        type=Path,
        default=DEFAULT_INPUT,
        help=f"Input image (default: {DEFAULT_INPUT})",
    )
    parser.add_argument(
        "--model",
        type=Path,
        default=DEFAULT_MODEL,
        help=f"MediaPipe Pose ONNX model (default: {DEFAULT_MODEL})",
    )
    parser.add_argument(
        "--output-dir",
        type=Path,
        default=PROJECT_DIR / "output",
        help="Directory for pose-image.jpg",
    )
    parser.add_argument(
        "--device",
        choices=("cpu", "cuda"),
        default="cpu",
        help="DNN execution device; CUDA requires a CUDA-enabled OpenCV build.",
    )
    parser.add_argument(
        "--score-threshold",
        type=float,
        default=0.5,
        help="Minimum landmark visibility and presence probability.",
    )
    display_group = parser.add_mutually_exclusive_group()
    display_group.add_argument(
        "--display",
        action="store_true",
        help="Open an interactive result window.",
    )
    display_group.add_argument(
        "--no-display",
        action="store_true",
        help="Run headlessly (the default; accepted explicitly for CI).",
    )
    parser.add_argument(
        "--validate",
        action="store_true",
        help="Check stable output invariants and print a success marker.",
    )
    return parser.parse_args(argv)


def run(args: argparse.Namespace) -> dict[str, object]:
    """Run the real image pipeline and return metrics used by regression tests."""

    input_path = args.input.expanduser().resolve()
    frame = cv2.imread(str(input_path), cv2.IMREAD_COLOR)
    if frame is None or frame.size == 0:
        raise FileNotFoundError(f"Could not read input image: {input_path}")
    if not 0.0 <= args.score_threshold <= 1.0:
        raise ValueError("--score-threshold must be between 0 and 1.")

    net = load_pose_model(args.model, args.device)
    result = infer_pose(net, frame)
    output, visible_count, edge_count = draw_pose(
        frame, result, args.score_threshold
    )
    output_path = args.output_dir.expanduser().resolve() / "pose-image.jpg"
    write_image(output_path, output)

    if args.validate:
        validate_pose(frame, result, visible_count, edge_count)
        print(
            "VALIDATION PASSED: "
            f"landmarks=33 visible={visible_count} edges={edge_count}"
        )

    if args.display and not args.no_display:
        cv2.imshow("MediaPipe Pose", output)
        cv2.waitKey(0)
        cv2.destroyAllWindows()

    print(f"OpenCV version: {cv2.__version__}")
    print(
        "POSE RESULT: "
        f"confidence={result.confidence:.6f} "
        f"visible={visible_count} edges={edge_count}"
    )
    print(f"Saved output: {output_path}")
    return {
        "output": output_path,
        "shape": output.shape,
        "confidence": result.confidence,
        "visible": visible_count,
        "edges": edge_count,
    }


def main(argv: list[str] | None = None) -> int:
    """Translate expected user errors into a concise nonzero CLI exit."""

    try:
        run(parse_args(argv))
    except (FileNotFoundError, RuntimeError, ValueError, cv2.error) as error:
        print(f"ERROR: {error}", file=sys.stderr)
        return 2
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
